### DISCUSSION TITLE
GitHub CI workflow windows build fails

### DISCUSSION THREAD

**Discussion by kmturley** [FluidSynth/fluidsynth#1140] [Q&A] [answered]
I am working on a cross-platform build of FluidSynth:
https://github.com/kmturley/fluidsynth/blob/master/.github/workflows/all.yml

Build works fine on Linux and Mac, but fails on Windows:
https://github.com/kmturley/fluidsynth/actions/runs/2858518817

The build error it fails on is:
```
-- Found PkgConfig: C:/ProgramData/chocolatey/bin/pkg-config.exe (found version "0.28") 
-- Checking for modules 'glib-2.0>=2.6.5;gthread-2.0>=2.6.5'
--   No package 'glib-2.0' found
--   No package 'gthread-2.0' found
CMake Error at C:/Program Files/CMake/share/cmake-3.24/Modules/FindPkgConfig.cmake:607 (message):
  A required package was not found
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.24/Modules/FindPkgConfig.cmake:829 (_pkg_check_modules_internal)
  CMakeLists.txt:481 (pkg_check_modules)
-- Configuring incomplete, errors occurred!

See also "D:/a/fluidsynth/fluidsynth/build/CMakeFiles/CMakeOutput.log".

See also "D:/a/fluidsynth/fluidsynth/build/CMakeFiles/CMakeError.log".
```

I see there is a windows build with commands here:
https://github.com/FluidSynth/fluidsynth/blob/master/.azure/azure-pipelines-win.yml

However it uses environment variables which are not in code, and also doesn't seem to have been run for awhile...Also pkg-config isntallation from choco installs from an old binary from sourceforge, which doesn't seem good.

How can I update my workflow to build FluidSynth for Windows on GitHub CI?
I've tried 35 different combinations of commands, and now feel like I might need some help :)

@derselbst was the last person working on that file.
@3119369616 might also know

**Comment by derselbst** [comment]
We have [two](https://github.com/FluidSynth/fluidsynth/blob/master/.azure/azure-pipelines-vcpkg.yml) Windows CI pipelines and comprehensive [docs in the wiki](https://github.com/FluidSynth/fluidsynth/wiki/BuildingWithCMake#building-with-mingw-on-windows). I'm pretty sure you'll figure it out. I'd recommend to use vcpkg as it's more up to date.

**Reply by kmturley** [reply]
I tried those approaches and:
- vcpkg approach is too slow, it takes 30 mins for each build, It's not possible to debug steps when each run takes that long
- I also tried replicating the windows only commands, but there are environment variables which are not in the yml file. For example `$(gtk-bundle)` and `$(libsndfile-url)`

As there are not recent runs I can't see any logs containing urls either

Neither url is mentioned in the documentation:
https://github.com/FluidSynth/fluidsynth/wiki/BuildingWithCMake

For gtk bundle there are several options on gtk.org, how do I know which is the right one?
```
gtk+-bundle_2.22.0-20101016_win64.zip | 24.4 MiB | 2010-Oct-16 17:03
gtk+-bundle_2.22.1-20101229_win64.zip | 24.6 MiB | 2010-Dec-29 14:13
gtk+-dev_2.22.0-1_win64.zip | 4.2 MiB | 2010-Oct-01 21:11
gtk+-dev_2.22.1-1_win64.zip | 4.2 MiB | 2010-Dec-29 13:37
gtk+_2.22.0-1_win64.zip | 8.7 MiB | 2010-Oct-01 21:15
gtk+_2.22.1-1_win64.zip
```
https://download.gnome.org/binaries/win64/gtk%2B/2.22/

libsndfile would I get from somewhere like this?
https://github.com/libsndfile/libsndfile/releases/tag/1.1.0

In addition this step is not going to work for me as I don't have access to those build artifacts, and wouldn't want to install artifacts I can't see the source of:
```
       buildType: specific
        # https://dev.azure.com/tommbrt/_apis/projects?api-version=5.0
        project: 'd3638885-de4a-4ce7-afe7-f237ae461c07'
        pipeline: 1
        artifactName: libinstpatch-XP-$(platform)
        downloadPath: '$(Build.ArtifactStagingDirectory)'
```

**Reply by kmturley** [reply]
I have found this version which appears to have all the urls configured:
https://github.com/netpipe/emscripten-projects/blob/f68ad1492f01629fdc2f3fd4ee1d096b76842d93/media/fluidsynth-emscripten/.appveyor.yml

Using this version I still get errors:
```
rm -rf C:/Strawberry/perl/bin/pkg-config*
mkdir /deps
cd /deps
curl -L "http://ftp.gnome.org/pub/gnome/binaries/win64/glib/2.26/glib_2.26.1-1_win64.zip" -o glib.zip
curl -L "http://ftp.gnome.org/pub/gnome/binaries/win64/glib/2.26/glib-dev_2.26.1-1_win64.zip" -o glib-dev.zip
curl -L "http://ftp.gnome.org/pub/gnome/binaries/win64/dependencies/pkg-config_0.23-2_win64.zip" -o pkg-config.zip
curl -L "http://ftp.gnome.org/pub/gnome/binaries/win64/dependencies/gettext-runtime_0.18.1.1-2_win64.zip" -o gettext.zip
curl -L "http://ftp.gnome.org/pub/gnome/binaries/win64/dependencies/proxy-libintl-dev_20100902_win64.zip" -o libintl-dev.zip
7z x glib.zip > NUL
7z x glib-dev.zip > NUL
7z x pkg-config.zip > NUL
7z x gettext.zip > NUL
7z x libintl-dev.zip > NUL
ls /deps/bin
export PATH=$PATH:/deps/bin
```

Has warnings for missing packages:
```
Checking for modules 'glib-2.0>=2.6.5;gthread-2.0>=2.6.5'
--   Found glib-2.0, version 2.26.1
--   Found gthread-2.0, version 2.26.1
-- Checking for module 'sndfile>=1.0.0'
--   No package 'sndfile' found
-- Checking for module 'libpulse-simple>=0.9.8'
--   No package 'libpulse-simple' found
-- Checking for module 'portaudio-2.0>=19'
--   No package 'portaudio-2.0' found
-- Checking for module 'jack'
--   No package 'jack' found
-- Checking for module 'libpipewire-0.3'
--   No package 'libpipewire-0.3' found
-- Checking for module 'dbus-1>=1.0.0'
--   No package 'dbus-1' found
-- Looking for ladspa.h
-- Looking for ladspa.h - not found
-- Checking for module 'libinstpatch-1.0>=1.1.0'
--   No package 'libinstpatch-1.0' found
-- Checking for module 'sdl2>=2.0.4'
--   No package 'sdl2' found
-- Could NOT find READLINE (missing: READLINE_LIBRARIES) 
-- Found OpenMP version: 2.0 date: 200203
--   OpenMP version is not supported. Feature disabled.
```

And then errors with:
```
-- Configuring done
CMake Error in src/CMakeLists.txt:
  Imported target "PkgConfig::GLIB" includes non-existent path

    "C:/Program"

```

**Reply by derselbst** [reply]
> vcpkg approach is too slow, it takes 30 mins for each build

One can use pipeline caching. Or docker containers... there are many ways around it.

> As there are not recent runs I can't see any logs containing urls either

There are plenty of recent runs available, all of them containing the urls. Just click the badges in the readme.

> For gtk bundle there are several options on gtk.org, how do I know which is the right one?

Usually, one should use the most recent one.


You seem to be inexperienced in compiling C programs on Windows. What is your motivation to port it to Github CI? Fluidsynth's Azure CI pipelines work fine... except for some randomly occurring HTTP timeouts, which I don't really care about. Why going through all the pain?

**Reply by kmturley** [reply]
1. Sure I can optimize later with caching or containers later, but to just get it to build in the first place it's too slow.
2. Ah I didn't realize jobs were running on https://dev.azure.com I'd been looking at: https://github.com/FluidSynth/fluidsynth/actions Using the Azure pipeline it is much easier to see the steps and the output. Thanks for the pointer!
3. Found the urls you use in the Azure pipeline output:
```
curl -LfsS -o gtk-bundle-dev.zip http://ftp.gnome.org/pub/gnome/binaries/win64/gtk+/2.22/gtk+-bundle_2.22.1-20101229_win64.zip   curl -LfsS -o libsndfile-dev.zip http://www.mega-nerd.com/libsndfile/files/libsndfile-1.0.28-w64.zip   || exit -1 
curl -LfsS -o mingw.zip https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/threads-win32/sjlj/x86_64-8.1.0-release-win32-sjlj-rt_v6-rev0.7z   || exit -1 
```

The libinstpatch binary is not publicly accessible:
```
Downloading artifact libinstpatch-x64 from: https://dev.azure.com/tommbrt//_apis/resources/Containers/19125121?itemPath=libinstpatch-x64&isShallow=true&api-version=4.1-preview.4
```
Does that mean I need to manually build a binary of the dependency from the source?
https://github.com/swami/libinstpatch

You are correct, I am inexperience in compiling C programs on Windows. I am however an experienced web developer at creating CI pipelines.

I create and maintain CI pipeline templates for plugin developers, using the main plugin frameworks:
https://github.com/studiorack/studiorack-workflows/tree/main/.github/workflows

My pipeline templates allow you to build any plugin for Mac, Linux and Win in a single pipeline, publishing binaries to GitHub releases in consistent format.

I am trying to build this popular plugin cross-platform:
https://github.com/Birch-san/juicysfplugin

However it fails when trying to compile FluidSynth (as a git submodule).
I need to to be able to build FluidSynth cross-platform to build JuicySF cross platform.

Why am I going through the pain? Because I like the challenge, to learn but also I believe I can share a template which will save others time in the future.

**Reply by kmturley** [reply]
I managed to get a successful cross-platform build using different commands on Windows:
```
    - name: Install Windows dependencies
      if: matrix.os == 'windows-latest'
      shell: cmd
      run: |
        @ECHO ON
        mkdir d:\deps || exit -1
        cd d:\deps || exit -1
        curl -LfsS -o gtk-bundle-dev.zip http://ftp.gnome.org/pub/gnome/binaries/win64/gtk+/2.22/gtk+-bundle_2.22.1-20101229_win64.zip || exit -1
        curl -LfsS -o libsndfile-dev.zip http://www.mega-nerd.com/libsndfile/files/libsndfile-1.0.28-w64.zip || exit -1
        7z x -aos -- gtk-bundle-dev.zip > NUL || exit -1
        7z x -aos -- libsndfile-dev.zip > NUL || exit -1
        REM need to fix the naming of libsndfile otherwise the linker won't find it
        mv lib\libsndfile-1.lib lib\sndfile.lib || exit -1
        mv lib\libsndfile-1.def lib\sndfile.def || exit -1
        DEL /F C:\Strawberry\perl\bin\pkg-config.bat

    - name: Build Windows
      if: matrix.os == 'windows-latest'
      shell: cmd
      run: |
        @ECHO ON
        SET "PATH=d:\deps\bin;%PATH%"
        mkdir build && cd build || exit -1
        cmake -Werror=dev -A x64 -T v142 -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=1 -DNO_GUI=1 -Dwindows-version=0x0A00 -Denable-jack=0 -Denable-pulseaudio=0 -Denable-ladspa=0 -Denable-dbus=0 -Denable-readline=0 -Denable-sdl2=0 -Denable-libinstpatch=0 .. || exit -1
        cmake --build . --config Release --parallel 3 || exit -1
        cmake --build . --config Release --target check || exit -1
        cmake --build . --config Release --target demo || exit -1
```

The main issues I encountered were:
- Using the correct dependency urls
- Using the correct shell: cmd instead of bash

Tomorrow I will look at converting the commands to bash and sharing commands between platforms!

**Reply by derselbst** [reply]
> Does that mean I need to manually build a binary of the dependency from the source? https://github.com/swami/libinstpatch

The windows binaries of libinstpatch are available from its CI pipeline in Azure DevOps. If you don't want fluidsynth to load DLS files, you don't need libinstpatch.


> I am trying to build this popular plugin cross-platform: https://github.com/Birch-san/juicysfplugin

But in this case I would recommend to better grab the windows binaries of fluidsynth's stable releases, rather than re-inventing a CI pipeline and potentially building unstable development versions. Also because you're now using very old dependency libraries and because those pipelines need maintenance.

**Reply by kmturley** [reply]
juicysfplugin could install fluidsynth as a binary, but wanted to make sure I understood how to build FluidSynth as a git submodule.
I went down the rabbit hole and won't be happy until I solve the problem!

**Comment by kmturley** [comment, marked as answer]
It took a while to navigate issues with cmd/bash, PATH variables, drive letters and missing dependencies. But I finally got it to work sharing the same commands cross-platform and building artifacts which are downloadable:
https://github.com/kmturley/fluidsynth/actions/runs/2893543767

```
name: FluidSynth All

defaults:
  run:
    shell: bash

on:
  workflow_dispatch:
  pull_request:
  push:

env:
  dir_build: build
  dir_source: .
  type: Release

jobs:
  build:
    name: All platforms
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        include:
          - os: ubuntu-latest
            name: linux
            generator: Unix Makefiles
            flags: -Denable-portaudio=1 -Denable-ladspa=1
            path: ./build/src
          - os: macos-latest
            name: mac
            generator: Unix Makefiles
            flags: -Denable-portaudio=1 -Denable-ladspa=1
            path: ./build/src
          - os: windows-latest
            name: win
            generator: Visual Studio 17 2022
            flags: -Denable-jack=0 -Denable-pulseaudio=0 -Denable-ladspa=0 -Denable-dbus=0 -Denable-readline=0 -Denable-sdl2=0 -Denable-libinstpatch=0
            path: ./build/src/Release

    steps:
    - uses: actions/checkout@v2

    - name: Install Linux dependencies
      if: matrix.os == 'ubuntu-latest'
      run: sudo apt-get update && sudo apt-get install cmake-data cmake libglib2.0-0 libsndfile-dev libasound2-dev libjack-dev portaudio19-dev libsdl2-dev libpulse-dev libdbus-1-dev libsystemd-dev libinstpatch-dev libreadline-dev g++-7 g++-8 clang-7 clang-8 clang-9 clang-10 clang-tidy-10

    - name: Install macOS dependencies
      if: matrix.os == 'macos-latest'
      run: brew install cmake pulseaudio portaudio jack dbus sdl2

    - name: Install Windows dependencies
      if: matrix.os == 'windows-latest'
      run: |
        mkdir D:/deps
        cd D:/deps
        curl -LfsS -o gtk-bundle-dev.zip http://ftp.gnome.org/pub/gnome/binaries/win64/gtk+/2.22/gtk+-bundle_2.22.1-20101229_win64.zip
        curl -LfsS -o libsndfile-dev.zip http://www.mega-nerd.com/libsndfile/files/libsndfile-1.0.28-w64.zip
        7z x -aos -- gtk-bundle-dev.zip
        7z x -aos -- libsndfile-dev.zip
        mv ./lib/libsndfile-1.lib ./lib/sndfile.lib
        mv ./lib/libsndfile-1.def ./lib/sndfile.def
        rm -rf C:/Strawberry/perl/bin/pkg-config*
        echo "/deps/bin" >> $GITHUB_PATH

    - name: Build
      run: |
        cmake \
          -G "${{ matrix.generator }}" \
          -DCMAKE_BUILD_TYPE="${{ env.type }}" \
          -DCMAKE_INSTALL_PREFIX=$HOME/fluidsynth_install \
          -DCMAKE_VERBOSE_MAKEFILE=1 \
          -DNO_GUI=1 \
          ${{ matrix.flags }} \
          -S "${{ env.dir_source }}" \
          -B "${{ env.dir_build }}"
        cmake --build ${{ env.dir_build }} --config ${{ env.type }} --parallel 3
        cmake --build ${{ env.dir_build }} --config ${{ env.type }} --target check
        cmake --build ${{ env.dir_build }} --config ${{ env.type }} --target demo
        cmake --build ${{ env.dir_build }} --config ${{ env.type }} --target install

    - name: List files generated
      run: |
        cd ${{ matrix.path }}
        find "$(pwd)"

    - uses: actions/upload-artifact@v3
      with:
        name: fluidsynth-${{ matrix.name }}
        path: ${{ matrix.path }}
```
